### ISSUE TITLE
HttpTest doesn't intercept calls when Flurl is used within ASP.NET Core 5.0 to call downstream services

### ISSUE THREAD

**Issue by agecas** [tmenier/Flurl#645]
HttpTest doesn't intercept calls when Flurl is used within ASP.NET Core 5.0 to call downstream services. I have found a few fixes for .NET Core 3.1 were setting **Server.PreserveExecutionContext = true;** fix the problem, however this does not seem to work on ASP.NET CORE 5.0

Here is repo with a sample: [https://github.com/agecas/FlurlIntegrationTestSample](https://github.com/agecas/FlurlIntegrationTestSample)

**Comment by tmenier** [comment]
I think the first thing to do is confirm beyond a doubt that it works as expected in ASP.NET Core 3.1. Can you target 3.1 with this confirm it?

**Comment by nreal** [comment]
This issue can you reproduce

**Comment by tmenier** [comment]
Can anyone confirm whether it works in ASP.NET Core 6? 5 is almost out of support already so I probably won't bother trying to fix it if it's only broken there. (I'm not certain a fix is even possible.) I'll close this in a few days if I don't hear from anyone but it can reopened if necessary.

**Comment by tmenier** [comment]
I downloaded your code and got to the bottom of this. The first problem is that where you're setting `Server.PreserveExecutionContext` (inside `FuturamaWebApplicationFactory.CreateServer`) never even gets called. You can prove it to yourself by putting a breakpoint inside that method and verifying it doesn't get hit. They also allude to this in the [docs](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.testing.webapplicationfactory-1.createserver):

> Creates the [TestServer](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.testhost.testserver?view=aspnetcore-6.0) with the bootstrapped application in builder. **This is only called for applications using [IWebHostBuilder](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.hosting.iwebhostbuilder?view=aspnetcore-6.0). Applications based on [IHostBuilder](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.ihostbuilder) will use [CreateHost(IHostBuilder)](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.testing.webapplicationfactory-1.createhost?view=aspnetcore-6.0#microsoft-aspnetcore-mvc-testing-webapplicationfactory-1-createhost(microsoft-extensions-hosting-ihostbuilder)) instead.**

I admit some of these things utterly confuse me but apparently `IWebHost` might get deprecated, the more generic `IHost` is the way forward, and it's what `WebApplicationFactory` works with by default. Since `CreateHost` is never called in this scenario, I have no idea where the "correct" place is to configure `TestServer`. But certainly the easiest place is directly in your test:

```c#
Factory.Server.PreserveExecutionContext = true;
var client = Factory.CreateClient();
...
```

**TL;DR That's the fix, but it still doesn't make your test pass. :)**

The above change caused your test to fail with a 500 error instead. That's because in your test setup you're telling Flurl to fake 405 response from sampleapios.com, not from your app. Inside your app, Flurl throws an exception (because 405 is a non-success code), which goes unhandled, resulting in a 500. For your test to pass you need to handle the 405 by catching `FlurlHttpException` or using `AllowAnyHttpStatus` and respond with a 405 from your app.

**Comment by tmenier** [comment]
Even better - if you want to keep it inside your custom factory, setting it in the constructor seemed to work like a charm:

```c#
public sealed class FuturamaWebApplicationFactory : WebApplicationFactory<Startup>
{
  public FuturamaWebApplicationFactory()
  {
    Server.PreserveExecutionContext = true;
  }
}
```
